### PR TITLE
[4764] Reactify Items Waiting For Approval Dashboard Widget

### PR DESCRIPTION
**Pull request by aachavarria**
https://github.com/craftercms/craftercms/issues/4764

### CODE REVIEW

File: ui/app/src/components/AwaitingApprovalDashletGrid/AwaitingApprovalDashletGridUI.tsx [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+                                <GlobalAppGridCell
+                                  className="width15"
+                                  title={new Intl.DateTimeFormat(
+                                    locale.localeCode,
+                                    locale.dateTimeFormatOptions
+                                  ).format(new Date(itemsLookup[path].live.lastPublishedDate))}
+                                >
+                                  {new Intl.DateTimeFormat(locale.localeCode, locale.dateTimeFormatOptions).format(
+                                    new Date(itemsLookup[path].live.lastPublishedDate)
+                                  )}
```
Comment (rart) [inline ui/app/src/components/AwaitingApprovalDashletGrid/AwaitingApprovalDashletGridUI.tsx]: Too many of these spread around. We should use a function (ask @jvega190 as we might have created it already) or try to use FormattedDate component if possible.

File: ui/app/src/components/AwaitingApprovalDashletGrid/AwaitingApprovalDashletGridUI.tsx [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+          <GlobalAppGridRow className="hoverDisabled">
+            <GlobalAppGridCell className="checkbox">
+              <Checkbox
+                indeterminate={isIndeterminate}
+                checked={isAllChecked || isIndeterminate}
+                onChange={onToggleCheckedAll}
+              />
+            </GlobalAppGridCell>
+            <GlobalAppGridCell className="width40 pl0">
+              <FormattedMessage id="dashboardItemsApproval.item" defaultMessage="Item" />
```
Comment (rart) [inline ui/app/src/components/AwaitingApprovalDashletGrid/AwaitingApprovalDashletGridUI.tsx]: Please update all instances of this id across files to awaitingApprovalDashlet

File: ui/app/src/components/DashboardApp/DashboardApp.tsx [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    const selected = Object.keys(selectedLookup).filter((path) => selectedLookup[path]);
+    if (selected.length === 0) {
+      return null;
+    } else if (selected.length) {
+      if (selected.length === 1) {
+        const path = selected[0];
+        const item = itemsByPath[path];
+        if (item) {
+          return generateSingleItemOptions(item, formatMessage, { includeOnly: actionsToBeShown }).flat();
+        }
```
Comment (rart) [inline ui/app/src/components/DashboardApp/DashboardApp.tsx]: Possible no return on this path

File: ui/app/src/components/DashboardApp/DashboardApp.tsx [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    }
+  }, [formatMessage, itemsByPath, selectedLookup]);
+
+  const fetchItems = (paths) => {
+    const filteredPaths = paths.filter((path) => !loadedRef.current.includes(path));
+    if (filteredPaths.length) {
+      fetchItemsByPath(site, filteredPaths, { preferContent: false }).subscribe((items) => {
+        setItemsByPath(createLookupTable(items, 'path'));
+        items.forEach((item) => {
+          loadedRef.current.push(item.path);
```
Comment (rart) [inline ui/app/src/components/DashboardApp/DashboardApp.tsx]: If there's any chance of items loading more than once, check for whether item is already in the collection.

File: ui/app/src/components/GlobalAppGridCell/GlobalAppGridCell.tsx [hunk 5]
```diff
@@ -36,8 +39,8 @@ const GlobalAppGridCell = withStyles((theme: Theme) =>
       '&.minWidth100': {
         minWidth: '100px'
       },
-      '&.bordered': {
-        borderBottom: `1px solid ${theme.palette.divider}`
+      '&.borderless': {
+        borderBottom: 0
```
Comment (rart) [inline ui/app/src/components/GlobalAppGridCell/GlobalAppGridCell.tsx:43]: Should be renamed (bb0 for example)

File: ui/app/src/styles/theme.tsx [hunk 6]
```diff
@@ -88,6 +88,11 @@ export const defaultThemeOptions: ThemeOptions = (function() {
         root: {
           display: 'block'
         }
+      },
+      MuiTableCell: {
+        sizeSmall: {
+          padding: '6px 10px'
```
Comment (rart) [inline ui/app/src/styles/theme.tsx:94]: What's the need to customise this?

File: ui/app/src/components/ItemStatesManagement/ItemStatesManagement.tsx [hunk 7]
```diff
@@ -228,7 +228,9 @@ export default function ItemStatesManagement(props: ItemStatesManagementProps) {
     } else {
       setItemStates(
         siteId,
-        Object.values(selectedItems).map((item) => item.path),
+        Object.values(selectedItems)
+          .filter((value) => value)
```
Comment (rart) [inline ui/app/src/components/ItemStatesManagement/ItemStatesManagement.tsx]:
```suggestion
          .filter(Boolean)
```

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/components/DashboardApp/DashboardApp.tsx (hunk 3 above)
Comment (aachavarria) [reply ui/app/src/components/DashboardApp/DashboardApp.tsx]: generateSingleItemOptions is returning [] when there are no options, and it should display a emptyMessage

File: ui/app/src/components/DashboardApp/DashboardApp.tsx (hunk 4 above)
Comment (aachavarria) [reply ui/app/src/components/DashboardApp/DashboardApp.tsx]: the items are only loaded one time

File: ui/app/src/components/DashboardApp/DashboardApp.tsx (hunk 3 above)
Comment (rart) [reply ui/app/src/components/DashboardApp/DashboardApp.tsx]: This whole chunk of code is the body of an `if`. What I'm saying is that, that `if` —above— is a code path which could lack a return (i.e. returns undefined) when item is nullish. Check if that's intended and managed.

File: ui/app/src/styles/theme.tsx (hunk 6 above)
Comment (aachavarria) [reply ui/app/src/styles/theme.tsx:94]: the default one is padding: 6px 24px 6px 16px, and make no sense to have right size to 24px, this is taking so much space and I think should be equals to left side

File: ui/app/src/components/DashboardApp/DashboardApp.tsx (hunk 3 above)
Comment (aachavarria) [reply ui/app/src/components/DashboardApp/DashboardApp.tsx]: ok, i will remove the if(item), the generateSingleItemOptions is already handle the item nullish

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 8]
```diff
@@ -385,11 +386,13 @@ function PublishDialogWrapper(props: PublishDialogProps) {
   const siteId = useActiveSiteId();
   const permissionsBySite = usePermissionsBySite();
   const myPermissions = permissionsBySite[siteId];
+  const dispatch = useDispatch();
 
   useUnmount(props.onClosed);
 
   const user = useSelector<GlobalState, GlobalState['user']>((state) => state.user);
   const submit = myPermissions.includes('publish') ? goLive : submitToGoLive;
+  const propagateAction = myPermissions.includes('publish') ? itemsApproved : itemsScheduled;
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx:395]: With the publish permission, you may still schedule. Please revisit this logic

File: ui/app/src/utils/datetime.ts [hunk 9]
```diff
@@ -50,3 +51,11 @@ export function asDayMonthDateTime(date: string): string {
     day: parts[4].value
   })} ${parts[6].value} @ ${getCurrentIntl().formatTime(date)}`;
 }
+
+export function asSimpleDateTime(
```
Comment (rart) [inline ui/app/src/utils/datetime.ts]:
```suggestion
export function asLocalizedDateTime(
```

File: ui/app/src/components/Dashlet/Dashlet.tsx [hunk 10]
```diff
@@ -21,24 +21,54 @@ import ExpandMoreIcon from '@material-ui/icons/ExpandMoreRounded';
...
+import RefreshRoundedIcon from '@material-ui/icons/RefreshRounded';
+import clsx from 'clsx';
 
 export type DashletProps = PropsWithChildren<
   Omit<AccordionProps, 'title'> & {
     title: React.ReactNode;
     headerRightSection?: React.ReactNode;
     icon?: ElementType;
+    onRefresh?(): void;
+    refreshDisabled?: boolean;
```
Comment (rart) [inline ui/app/src/components/Dashlet/Dashlet.tsx:36]: **When** do you need to disable the refresh button? Should rather the whole interface be disabled?

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx (hunk 8 above)
Comment (aachavarria) [reply ui/app/src/modules/Content/Publish/PublishDialog.tsx:395]: will check this on other ticket, this behavior is not part of this ticket

File: ui/app/src/components/Dashlet/Dashlet.tsx (hunk 10 above)
Comment (aachavarria) [reply ui/app/src/components/Dashlet/Dashlet.tsx:36]: when there is an in progress action on the body, like an Fetch call, and the ui should be clickable to collapse or expand

**Review by rart — APPROVED** [review]

**Pull request merged**